torch/ao/quantization/fx/utils.py

```diff
@@ -640,3 +641,220 @@ def get_skipped_module_name_and_classes(
+                 v            v
+                 lstm_output_dq
+                       |
+                       v
+                original_user(s)
+
+    For step (4), reroute all users of the original LSTM node(s) as follows:
+
+      lstm_output -> lstm_output_dq
+      lstm_output[0] -> output_dq
+      lstm_output[1] -> hidden_dq
+      lstm_output[1][0] -> hidden0_dq
+      lstm_output[1][1] -> hidden1_dq
+
+    Return the node `lstm_output_dq`.
+    """
+    # (1) Split the LSTM node into (output, (hidden0, hidden1))
+    # (2) Insert a DeQuantStub after each internal node
+    original_lstm_users = list(node.users.keys())
+    with graph.inserting_after(node):
+        output = graph.call_function(operator.getitem, (node, 0))
+        output_dq = _insert_dequant_stub(output, model, named_modules, graph)
+    with graph.inserting_after(output_dq):
+        hidden = graph.call_function(operator.getitem, (node, 1))
+    with graph.inserting_after(hidden):
+        hidden0 = graph.call_function(operator.getitem, (hidden, 0))
+        hidden0_dq = _insert_dequant_stub(hidden0, model, named_modules, graph)
+    with graph.inserting_after(hidden0_dq):
+        hidden1 = graph.call_function(operator.getitem, (hidden, 1))
+        hidden1_dq = _insert_dequant_stub(hidden1, model, named_modules, graph)
+
+    # (3) Recombine the DeQuantStubs into the same structure as before
+    with graph.inserting_after(hidden1_dq):
+        hidden_dq = graph.call_function(tuple, ([hidden0_dq, hidden1_dq],))
+    with graph.inserting_after(hidden_dq):
+        lstm_output_dq = graph.call_function(tuple, ([output_dq, hidden_dq],))
+
+    # (4) Reroute all consumers of the original LSTM node
+    # First, gather the original LSTM nodes
+    node_to_replacement = {node: lstm_output_dq}
+    for user in list(node.users.keys()):
+        if _is_getitem_node(user) and user.args[1] == 0:
+            node_to_replacement[user] = output_dq  # lstm_output[0]
+        if _is_getitem_node(user) and user.args[1] == 1:
+            node_to_replacement[user] = hidden_dq  # lstm_output[1]
+            for hidden_user in list(user.users.keys()):
+                if _is_getitem_node(hidden_user) and hidden_user.args[1] == 0:
+                    node_to_replacement[hidden_user] = hidden0_dq  # lstm_output[1][0]
+                if _is_getitem_node(hidden_user) and hidden_user.args[1] == 1:
+                    node_to_replacement[hidden_user] = hidden1_dq  # lstm_output[1][1]
+    # Do not reroute nodes we added or the original nodes that will be consumed
+    nodes_added = [output, output_dq, hidden, hidden_dq, hidden0, hidden0_dq, hidden1, hidden1_dq]
+    nodes_to_skip = set(node_to_replacement.keys()).union(nodes_added)
+    for original_node, replacement in node_to_replacement.items():
+        for user in list(original_node.users.keys()):
+            if user not in nodes_to_skip:
+                user.replace_input_with(original_node, replacement)
```

Comment:
do we need to do this? is it enough to just update all uses of `lstm_output_dq`?

Reply:
I guess it might break the reference quantized pattern matching, but feels like we should update the matching code instead

Reply:
or have some optimization passes, e.g.:
```
x --> c = tuple(x, ...) -> c[0] --> output
```
--->

```
x --> output
```

Reply:
Yes, it would break the reference pattern matching. Also, redirecting just `lstm_output_dq` would add some unnecessary nodes for simple use cases like `lstm[0]`. I feel changing the matching logic is a much bigger scope than this task. Maybe we should do it separately?

Reply:
how about this suggestion: https://github.com/pytorch/pytorch/pull/85068#discussion_r975889634

Reply:
As discussed offline, we will have an additional pass in this PR to match the pattern `tuple -> tuple -> getitem -> getitem`, for an arbitrary number of `tuple/getitem` pairs